Add podScaleUpDelay annotation support

damirda opened the pull request:
#### Which component this PR applies to?

cluster-autoscaler

#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Extends existing option/feature `new-pod-scale-up-delay` that is global for all the pods and adds ability to specify custom delay per pod through annotation `cluster-autoscaler.kubernetes.io/pod-scale-up-delay`. This allows better control over scale up process in the situations when we are running a lot of batch jobs at the regular intervals, like _run this job every hour, but it is ok to wait 5 minutes to start_. 

#### Which issue(s) this PR fixes:

There was a discussion on this issue and someone suggested this solution. I liked it, implemented it, and it worked well for what I needed.
https://github.com/kubernetes/autoscaler/issues/3916

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
```release-note
A new annotation `cluster-autoscaler.kubernetes.io/pod-scale-up-delay` has been added, which allows user to set `new-pod-scale-up-delay` per pod, instead of autoscaler wide configuration. Annotation value will be used only if it is larger then global `new-pod-scale-up-delay`.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

x13n commented on cluster-autoscaler/core/static_autoscaler.go:
```diff
@@ -719,13 +722,27 @@ func (a *StaticAutoscaler) nodeGroupsById() map[string]cloudprovider.NodeGroup {
+		if podScaleUpDelayAnnotationStr, ok := pod.Annotations[podScaleUpDelayAnnotationKey]; ok {
+			podScaleUpDelayAnnotation, err := time.ParseDuration(podScaleUpDelayAnnotationStr)
+			if err != nil {
+				klog.Errorf("Failed to parse pod annotation: %v", err)
```
nit: Can you identify the pod and annotation in the logs? Maybe `klog.Errorf("Failed to parse pod %q annotation %s: %v", pod.Name, podScaleUpDelayAnnotationKey, err)`

x13n reviewed (commented):
The code change LGTM, but I think we also need:
- a unit test confirming this works 
- FAQ entry about the new annotation

Can you add it?

damirda commented:
@x13n Thank you for your help with this PR so far!
I added FAQ entry that basically covers my use case. I also added unit test for the existing function that I extended.
Let me know if I should add more tests or if you think FAQ update should be more generalized.

x13n commented on cluster-autoscaler/FAQ.md:
```diff
@@ -331,6 +332,23 @@ CA might scale down non-empty nodes with utilization below a threshold
+### How can I prevent Cluster Autoscaler from scaling up too quickly?
+
+Running many shortlived scheduled jobs in a regular intervals could lead to cluster running in
+cycle of scaling up, executing jobs quickly and then waiting for `scale-down-unneeded-time`
```
Increasing `--scale-down-unneeded-time` is also a way of improving the situation. In a way the two flags are complementary: `--scale-down-unneeded-time` controls how long a node will wait for pods, `--new-pod-scale-up-delay` controls how long a pod will wait for a node.

damirda replied:
How would increasing `--scale-down-unneeded-time` improve the situation? You will just get nodes that are started running even longer unneeded?

x13n replied:
The way I understand the problem here is that there is too high node churn - too many nodes coming down and up again, wasting compute resources on startup/shutdown. Keeping nodes for a longer time after they become empty is a way of addressing this, so is introducing a delay before triggering a scale up.

If the problem is phrased as "scale up happens too quickly" then it can also help: nodes are not deleted, giving future pods a chance to get scheduled on them instead of triggering a scale up.

damirda replied:
My problem was that nodes are coming up too fast and then staying up too long. :smile: 
We did decrease `--scale-down-unneeded-time` to 5 minutes and it did help, but anyway, I understand your point. I will write in a way to cover different cases, depending on what is that you want to achieve.

x13n commented on cluster-autoscaler/FAQ.md:
```diff
@@ -621,7 +639,8 @@ then this node group may be excluded from future scale-ups.
 By default, scale-up is considered up to 10 seconds after pod is marked as unschedulable, and scale-down 10 minutes after a node becomes unneeded.
```
Can we avoid duplicating content? I'd drop most of this paragraph and leave just:

> By default, scale-up is considered up to 10 seconds after pod is marked as unschedulable, and scale-down 10 minutes after a node becomes unneeded. Read [this section](...) to see how you can modify this behavior.

damirda commented on cluster-autoscaler/core/static_autoscaler.go:
```diff
@@ -719,17 +722,32 @@ func (a *StaticAutoscaler) nodeGroupsById() map[string]cloudprovider.NodeGroup {
 	for _, pod := range allUnschedulablePods {
 		podAge := currentTime.Sub(pod.CreationTimestamp.Time)
-		if podAge > newPodScaleUpDelay {
+		// read podScaleUpDelay from annotations
```
Yes, I agree that comment is unnecessary.

damirda commented:
This failing test seem to be flaky. Works fine locally and also `master` is not broken.
Plus completely unrelated to my changes. What is the best way to handle that?

damirda commented:
Never mind, restarting the build helped...

x13n commented on cluster-autoscaler/core/static_autoscaler_test.go:
```diff
@@ -1377,6 +1380,103 @@ func TestSubtractNodes(t *testing.T) {
+			newPodScaleUpDelay: 0,
+			runTime:            now,
+			pods:               []*apiv1.Pod{p1, p4},
+			expectedPods:       []*apiv1.Pod{p1},
```
I don't think pods with malformed annotation should be ignored. I think logging an error should be enough - if annotation cannot be parsed, CA behavior should fall back to global setting instead of banning the pod forever.

damirda replied:
There was a bug in my tests, so it didn't catch that my code reformatting commit actually broke this behavior. Now I reverted code reformatting (pretty much completely) and fixed the tests.

x13n commented on cluster-autoscaler/core/static_autoscaler_test.go:
```diff
@@ -1377,6 +1380,103 @@ func TestSubtractNodes(t *testing.T) {
+			actual := autoscaler.filterOutYoungPods(tt.pods, tt.runTime)
+
+			assert.ObjectsAreEqual(tt.expectedPods, actual)
+			if tt.expectedError != "" {
```
nit: Is this condition changing anything? `Contains(t, buf.String(), "")` is always true, isn't it?

damirda replied:
I just think it makes test more readable - if there is an error expected, check if we got it. But it would work wothout `if` as well.

x13n commented on cluster-autoscaler/core/static_autoscaler_test.go:
```diff
@@ -1430,22 +1430,22 @@ func TestFilterOutYoungPods(t *testing.T) {
 			name:               "annotation delay smaller than global",
 			newPodScaleUpDelay: 5 * time.Minute,
 			runTime:            now,
 			pods:               []*apiv1.Pod{p1, p3},
```
`p3` has annotation of 2m and is created 1m before `runTime`, so it wouldn't be included even if the implementation allowed for annotation-based delay to be smaller than the global one. I think this test should use `runTime: now.Add(2 * time.Minute)` to actually show the difference.

damirda replied:
Yes, that is actually very good point. Thank you!

x13n reviewed (commented):
LGTM, just one suggested improvement in tests.

x13n commented:
/lgtm

damirda commented:
@x13n Thank you for all your help on this PR.
I will squash all the commits in the end and prepare this for a merge.

damirda commented:
/cc @aleksandra-malinowska @feiskyer

x13n commented:
Merge will happen automatically once this has both lgtm and approved labels, so I'd suggest doing it first (it will also reset the lgtm tag, so I'll need to reapply it).

x13n commented:
/lgtm

x13n commented:
/assign @towca

damirda commented:
@towca Could we get a review here?

towca commented:
Looks good to me, thanks for the contribution!

/approve
